File: app/src/main/java/com/example/myapplication/ui/answer/AnswerViewModel.kt
package com.example.myapplication.ui.answer

import androidx.compose.runtime.Immutable
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.snaplearn.data.QuestionAnswer
import com.example.snaplearn.viewmodel.AnswerState
import com.example.snaplearn.viewmodel.SharedAnswerViewModel
import com.example.snaplearn.viewmodel.SharedHistoryViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Android-specific ViewModel for the Answer screen that adapts the shared
 * AnswerViewModel for Android UI.
 */
class AnswerViewModel(
    private val sharedAnswerViewModel: SharedAnswerViewModel,
    private val sharedHistoryViewModel: SharedHistoryViewModel
) : ViewModel() {
    
    private val _uiState = MutableStateFlow<AnswerUiState>(AnswerUiState.Initial)
    val uiState: StateFlow<AnswerUiState> = _uiState.asStateFlow()
    
    private var lastQuestion: String = ""
    private var lastAnswer: String = ""
    
    init {
        // Observe the shared ViewModel state
        viewModelScope.launch {
            sharedAnswerViewModel.state.collect { state ->
                when (state) {
                    is AnswerState.Initial -> {
                        _uiState.value = AnswerUiState.Initial
                    }
                    is AnswerState.Loading -> {
                        _uiState.value = AnswerUiState.Loading
                    }
                    is AnswerState.Success -> {
                        lastAnswer = state.text
                        _uiState.value = AnswerUiState.Success(state.text)
                    }
                    is AnswerState.Error -> {
                        _uiState.value = AnswerUiState.Error(state.message)
                    }
                }
            }
        }
    }
    
    /**
     * Generates an answer for the given question.
     */
    fun generateAnswer(question: String) {
        lastQuestion = question
        sharedAnswerViewModel.generateAnswer(question)
    }
    
    /**
     * Saves the current question and answer to history.
     */
    fun saveToHistory() {
        if (lastQuestion.isNotBlank() && lastAnswer.isNotBlank()) {
            viewModelScope.launch {
                sharedHistoryViewModel.saveQuestionAnswer(lastQuestion, lastAnswer)
                _uiState.update { 
                    if (it is AnswerUiState.Success) {
                        AnswerUiState.Success(it.text, true)
                    } else {
                        it
                    }
                }
            }
        }
    }
}

/**
 * UI state for the Answer screen.
 */
@Immutable
sealed class AnswerUiState {
    /**
     * Initial state, before any generation has been attempted.
     */
    object Initial : AnswerUiState()
    
    /**
     * Loading state, when generation is in progress.
     */
    object Loading : AnswerUiState()
    
    /**
     * Success state, containing the generated answer.
     */
    data class Success(val text: String, val saved: Boolean = false) : AnswerUiState()
    
    /**
     * Error state, containing the error message.
     */
    data class Error(val message: String) : AnswerUiState()
} 

File: iosApp/iosApp/AnswerViewModel.swift
import Foundation
import shared
import Combine

/// The different states of the answer screen.
enum AnswerViewState {
    case initial
    case loading
    case success(text: String, saved: Bool = false)
    case error(message: String)
}

/// ViewModel for the Answer screen.
@MainActor
class AnswerViewModel: ObservableObject {
    @Published var state: AnswerViewState = .initial
    
    private let sharedViewModel: SharedAnswerViewModel
    private let historyViewModel: SharedHistoryViewModel
    private var lastQuestion: String = ""
    private var lastAnswer: String = ""
    private var stateObservationTask: Task<Void, Never>? = nil
    
    init(sharedViewModel: SharedAnswerViewModel, historyViewModel: SharedHistoryViewModel) {
        self.sharedViewModel = sharedViewModel
        self.historyViewModel = historyViewModel
        
        startObservingState()
    }
    
    deinit {
        stateObservationTask?.cancel()
    }
    
    /// Start observing the shared ViewModel state.
    private func startObservingState() {
        stateObservationTask?.cancel()
        
        stateObservationTask = Task {
            for await state in sharedViewModel.state {
                if let initialState = state as? AnswerState.Initial {
                    self.state = .initial
                } else if let loadingState = state as? AnswerState.Loading {
                    self.state = .loading
                } else if let successState = state as? AnswerState.Success {
                    self.lastAnswer = successState.text
                    if case let .success(_, saved) = self.state {
                        self.state = .success(text: successState.text, saved: saved)
                    } else {
                        self.state = .success(text: successState.text)
                    }
                } else if let errorState = state as? AnswerState.Error {
                    self.state = .error(message: errorState.message)
                }
            }
        }
    }
    
    /// Generate an answer for the given question.
    func generateAnswer(question: String) {
        lastQuestion = question
        sharedViewModel.generateAnswer(prompt: question)
    }
    
    /// Save the current question and answer to history.
    func saveToHistory() {
        if !lastQuestion.isEmpty && !lastAnswer.isEmpty {
            Task {
                historyViewModel.saveQuestionAnswer(question: lastQuestion, answer: lastAnswer)
                if case let .success(text, _) = state {
                    state = .success(text: text, saved: true)
                }
            }
        }
    }
} 